File: frontend/src/app/models/recipe.ts
import { Difficulty } from './difficulty';

export interface RecipeStep {
  id: number;
  stepNumber: number;
  instruction: string;
}

export interface RecipeSection {
  id: number;
  title: string;
  sortOrder: number;
  steps: RecipeStep[];
}

export interface IngredientLine {
  id: number;
  ingredientId: number;
  ingredientName: string;
  amount: number;
  unit: string;
}

export interface CategoryRef {
  id: number;
  name: string;
}

export interface Recipe {
  id: number;
  title: string;
  description?: string;
  prepTimeMinutes: number;
  cookTimeMinutes: number;
  servings: number;
  difficulty: Difficulty;
  imageUrl?: string;
  authorId: string;
  createdAt: string;
  updatedAt: string;
  sections: RecipeSection[];
  ingredients: IngredientLine[];
  categories: CategoryRef[];
  averageRating?: number;
  ratingCount: number;
}

export interface StepRequest {
  stepNumber: number;
  instruction: string;
}

export interface SectionRequest {
  title: string;
  sortOrder: number;
  steps: StepRequest[];
}

export interface IngredientLineRequest {
  ingredientId: number;
  amount: number;
  unit: string;
}

export interface RecipeRequest {
  title: string;
  description?: string;
  prepTimeMinutes: number;
  cookTimeMinutes: number;
  servings: number;
  difficulty: Difficulty;
  imageUrl?: string;
  sections: SectionRequest[];
  ingredients: IngredientLineRequest[];
  categoryIds?: number[];
}

File: backend/src/main/java/ch/uscata/alex/recipevault/recipe/Recipe.java
package ch.uscata.alex.recipevault.recipe;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import ch.uscata.alex.recipevault.category.Category;
import ch.uscata.alex.recipevault.rating.Rating;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "recipes")
@Getter
@Setter
@NoArgsConstructor
public class Recipe {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 150)
    private String title;

    @Column(length = 2000)
    private String description;

    @Column(name = "prep_time_minutes", nullable = false)
    private Integer prepTimeMinutes;

    @Column(name = "cook_time_minutes", nullable = false)
    private Integer cookTimeMinutes;

    @Column(nullable = false)
    private Integer servings;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 10)
    private Difficulty difficulty;

    @Column(name = "image_url", length = 500)
    private String imageUrl;

    @Column(name = "author_id", nullable = false, updatable = false)
    private UUID authorId;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("sortOrder ASC")
    private List<RecipeSection> sections = new ArrayList<>();

    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RecipeIngredient> ingredients = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "recipe_categories",
            joinColumns = @JoinColumn(name = "recipe_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private Set<Category> categories = new HashSet<>();

    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Rating> ratings = new ArrayList<>();
}
